sdk/identity/Azure.Identity/src/ImdsManagedIdentityProbeSource.cs

```diff
@@ -0,0 +1,195 @@
+            {
+                if (e.InnerException is TaskCanceledException)
+                {
+                    throw;
+                }
+                else
+                {
+                    throw new CredentialUnavailableException(NoResponseError, e);
+                }
+            }
+            catch (TaskCanceledException e)
+            {
+                throw new CredentialUnavailableException(NoResponseError, e);
+            }
+            catch (AggregateException e)
+            {
+                throw new CredentialUnavailableException(AggregateError, e);
+            }
+            catch (CredentialUnavailableException)
+            {
+                throw;
+            }
+            catch (ProbeRequestResponseException)
+            {
+                // This was an expected response from the IMDS endpoint without the Metadata header set.
+                // Re-issue the request (CreateRequest will add the appropriate header).
+#pragma warning disable AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+                var authResult = await _client.AcquireTokenForManagedIdentityAsync(context, cancellationToken).ConfigureAwait(false);
+                return authResult.ToAccessToken();
+#pragma warning restore AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+            }
+        }
+
+        protected override async ValueTask<AccessToken> HandleResponseAsync(bool async, TokenRequestContext context, HttpMessage message, CancellationToken cancellationToken)
+        {
+            Response response = message.Response;
+            // if we got a response from IMDS we can stop limiting the network timeout
+            _imdsNetworkTimeout = null;
+
+            // handle error status codes indicating managed identity is not available
+            string baseMessage = response.Status switch
+            {
+                400 when IsProbRequest(message) => throw new ProbeRequestResponseException(),
+                400 => IdentityUnavailableError,
+                502 => GatewayError,
+                504 => GatewayError,
+                _ => default
+            };
+
+            if (baseMessage != null)
+            {
+                string content = new RequestFailedException(response, null, new ImdsRequestFailedDetailsParser(baseMessage)).Message;
+
+                var errorContentMessage = await GetMessageFromResponse(response, async, cancellationToken).ConfigureAwait(false);
+
+                if (errorContentMessage != null)
+                {
+                    content = content + Environment.NewLine + errorContentMessage;
+                }
+
+                throw new CredentialUnavailableException(content);
+            }
+
+            var token = await base.HandleResponseAsync(async, context, message, cancellationToken).ConfigureAwait(false);
+            return token;
+        }
+
+        public static bool IsProbRequest(HttpMessage message)
```

Comment:
```suggestion
        public static bool IsProbeRequest(HttpMessage message)
```

nit:  I think that's a typo?

Reply:
yes, it is :) It's internal, but still a typo. Will fix